Email alert notification receipts

gpeng opened the pull request:
This work is part of the ongoing efforts to improve travel advice alert email monitoring and will allow us to reconcile a 'received' email with a published edition in Travel Advice Publisher (where we'll add checks that can be run by Jenkins -> Icinga)

This project sets up an S3 bucket and a lambda function that parses files and renames them with a prefix (currently only `travel-advice-alerts`) and the `govuk_request_id` parsed from the email body.

An SES domain and RuleSet and an S3 event source are also required but can't currently be configured via Terraform. (details in the README)

I've not been able to test this via the `rake` tasks from the repo root as I've not got access to the state files, environments etc so I'm not sure that the paths work from there.

The lambda code is deployed from a zip file which I've checked in and included in this PR. I'm not sure this is the best approach as it's 'unreviewable'. I'd welcome any better ideas. Zipping the file as part of the rake task to apply maybe?

Apologies for the one big commit. Code was moved from a local repo. Happy to break up if it will help with reviews.

UPDATE: Added an `aws_s3_bucket_notification` to trigger the lambda. This is only supported in Terraform 0.6.15

surminus commented on projects/email_alert_notifications/email_alert_notifications.tf:
```diff
@@ -0,0 +1,55 @@
+provider "aws" {}
```
This isn't required because [it is supplied as part of common variables](https://github.com/alphagov/govuk-terraform-provisioning/blob/master/configs/common_variables.tf)

surminus commented:
I _think_ currently the way the Rakefile deploys stuff is to copy everything from resource directories with the *.tf filename extension into a temporary directory and then deploy, so this may miss out the zip/py files, so we'll need to tweak that. I think it's also worth discussing whether we want "extra" files within the resource/ dir. 

Currently if an environment isn't defined within the resources/ dir then it'll assume that you want it deployed to all environments.

deanwilson commented:
The file paths that use `path.module` will all be incorrect I think, as this isn't a module. You can use relative ones although I suspect something like path.cwd might also work.

deanwilson commented:
This depends on #47 being merged.

boffbowsh commented on projects/email_alert_notifications/templates/email_alert_s3_bucket_policy.json:
```diff
@@ -0,0 +1,32 @@
+{
+  "Version": "2012-10-17",
+  "Id": "Policy1460992104238",
```
Could give this a proper ID I think?

gpeng replied:
Fixed in ca7c3c1. Seems that using a `uuid` is the way to go although I haven't used an `ID` at all in the other policies. I'm not sure if there's any benefit to using them. As it happens this policy confuses `terraform apply` into thinking it has changed every time it is run which I hoped the change you suggested might fix but didn't unfortunately. I think it's line 27 not being specific enough but haven't been able to work out a more 'fully qualified' version that works.

deanwilson commented:
In terms of the lambda build I've been looking at something like this rather than building the zip manually.

    resource "null_resource" "build_poller_zip" {
        provisioner "local-exec" {
            command = "zip -j py-poller.zip files/url-poller.py"
        }
    }

    resource "aws_lambda_function"
        depends_on = [ "null_resource.build_poller_zip"]

Would you be willing to give this a go on this PR?

surminus commented:
We're going to add some code that works with the additional files, but the structure needs to be rejigged:

- *.tf files should live in `<project>/resources`
- templates can go in `<project>/templates`
- any other files should go in `<project>/files`

The README can stay in the project root. If you remove the references to the `module.path` then our rake tasks should hopefully do the right thing.

radeksimko commented on projects/email_alert_notifications/resources/email_alert_notifications.tf:
```diff
@@ -0,0 +1,63 @@
+  role = "${aws_iam_role.lambda_execute_and_write_to_email_alert_bucket.arn}"
+  handler = "rename_email_files_with_request_id.lambda_handler"
+  runtime = "python2.7"
+  source_code_hash = "${base64sha256(file("files/rename_email_files_with_request_id.py"))}"
```
FYI: `source_code_hash` expects a hash of the ZIP file, so you'll keep getting unnecessary diffs in `plan` with the current approach as the hash will never match with what the API returns.

gpeng replied:
Thanks for your comment. I was using the zip file path originally but in an effort to try and avoid having zipped files in the repo I've moved its creation into a `null_resource`. As a result the `base64sha256` function raises a `no such file` error if I use the zip file. Is there a better way that I can trigger this resource to be updated when the .py source changes?

radeksimko replied:
> Is there a better way that I can trigger this resource to be updated when the .py source changes?

Before Terraform gets to `plan/apply` stage, it (by default) performs `refresh` (effectively `Read` from CRUD for each resource in all `*.tf` files) so it can figure out what may have changed and how does it affect `plan/apply`. After refresh is done for all resources, it performs `plan/apply`.

> As a result the base64sha256 function raises a no such file error if I use the zip file.

The `base64sha256()` function is executed at the time of `refresh` since it has static value (static path to zip file) hence Terraform won't mark it as "computed". All provisioners are always executed at the time of `apply`, not during `refresh`. That's why you're hitting this.

To mark it as "computed" you would probably need to have a 1st class support for "zipping" in terraform which would be returning the shasum that you would reference in `aws_lambda_function. source_code_hash`. There are some (good) reasons why Terraform doesn't support this (yet):
https://github.com/hashicorp/terraform/pull/3858

The only way to get around this would be possibly executing terraform twice (+ removing any unnecessary `triggers` & `depends_on`):

```
terraform apply -target=null_resource.build_rename_email_files_with_request_id
terraform plan/apply
```

that way terraform won't perform any validation/checks on any resources except the targeted one.
Obviously the `null_resource`'s provisioner would be executed twice (in the second run too), but at least Terraform would get a chance to prepare the right `plan` and `apply` it correctly.

Also keep in mind that any `local-exec` provisioner makes your Terraform files platform-specific - i.e. the requirement suddenly isn't just terraform binary, but also platform-specific `zip` utility having the same interface (e.g. supporting `-j` flag). This is also why I think that building ZIP files via CI is a better solution.

gpeng commented:
@surminus I've reorganised the files as you suggested. I think it will need a slight change to #47 as per my comments there

@deanwilson I've used that `null_resource` approach you suggested with a couple of tweaks to get it running. Good shout :+1:

radeksimko commented:
I'm not entirely sure how you do deal with other artefacts, like ruby gems in GDS, but my friendly recommendation would be to treat the ZIP files for Lambda functions the same way as other artefacts - i.e. have their own repo with CI (Jenkins/Travis/Circle) that is running tests (if you have any) and most importantly generating ZIP file and uploading it into a versioned S3 bucket. That build in your CI can spit out an S3 `VersionId` of the file it has just uploaded.

Then you can just bump the `s3_object_version` in `*.tf` and actually _release_ the ZIP file. It may look like too many steps, but it's codebase like any other and IMO should be treated the same way.

There is also a separate thread about proper [versioning](https://github.com/hashicorp/terraform/issues/6067) of Lambda Functions, which isn't well supported in Terraform yet.

radeksimko commented on projects/email_alert_notifications/README.md:
```diff
@@ -0,0 +1,32 @@
+* An S3 event configured to execute the lambda (`rename_email_files_with_request_id`)
+on the `ObjectCreated` `Put` event for the bucket
+
+Terraform v0.6.14 doesn't support S3 event sources but support is in
```
`0.6.15` was released just a few hours ago 😉 
https://www.terraform.io/downloads.html

gpeng commented:
Thanks for all your comments @radeksimko . Really helpful.

radeksimko commented:
There is an ongoing effort to release Terraform 0.7 soon which will contain something that might help you to make the workflow cleaner/more automated - Data Sources.

Effectively you'd be able to do something like this:

```hcl
data "aws_s3_object" "lambda_zip_file" {
  bucket = "my-lambda-functions"
  path = "lambda.zip"
}
...
resource "aws_lambda_function" "rename_email_files_with_request_id" {
  s3_object_version = "${data.aws_s3_object.lambda_zip_file.latest_version}"
```
That way you wouldn't need that extra step of bumping Version ID manually, but you'd just let Terraform to pull down the latest from S3 at the time of `refresh`.

The syntax and actual implementation is [WIP](https://github.com/hashicorp/terraform/pull/4961) though - this is just to give you an idea.

gpeng commented:
I've updated the directory structure and reverted the `null_resource` commit. Once we've settled on where the zipped lambdas are going to live I'll add support for that.

surminus commented:
@gpeng as suggested in https://github.com/alphagov/govuk-terraform-provisioning/pull/47#discussion_r62853746 I think what I stated above was incorrect.

gpeng commented:
Thanks... I've laid it out as per #47 so I think it should be good to go now. Just need to update to pull the lambda zip from s3 once that is setup.

surminus commented:
We have the config in for buckets to store the lambda zips from (https://github.com/alphagov/govuk-terraform-provisioning/pull/50) although not sure if they're deployed yet (easy to do). I guess the next thing to do is https://trello.com/c/eukUJfOU/84-write-jenkins-deployment-job-for-lambda-apps

surminus commented:
@gpeng, we've just had a chat about this. To deploy this the process would be:
1) Dev team add a `jenkins.sh` script which zips up the code for the lambda function
2) Jenkins uploads the zip to an S3 bucket
3) Jenkins kicks off this Terraform job to deploy the function

We are prioritising steps 2 & 3 along with setting up the email endpoints.

gpeng commented:
Ok... Sounds good. So would we have a convention for where zipped lambdas live within a typical app and Jenkins would upload any that it finds in the 'standard' location `/lambdas/<whatevs.zip>`?

surminus commented:
That sounds like a good plan.

gpeng commented:
Would that mean that we'd have to have a separate update/install lambda Terraform job?

surminus commented:
I think this is fine as it is for the time-being and don't think we should have a separate job. This is dependent on https://github.com/alphagov/govuk-terraform-provisioning/pull/47 which I think @deanwilson has thoughts on, though I am happy with it in this way and don't really mind either way. I _think_ this terraform project looks fine to run on each deployment (though if you want to do the work to split it up feel free).

[For this part](https://github.com/alphagov/govuk-terraform-provisioning/pull/46/files#diff-e4523e052853d909202ff683d5b738bbR44), will the source change as the `lambda.zip` will be stored in another S3 bucket [that were created here](https://github.com/alphagov/govuk-terraform-provisioning/pull/50)?

gpeng commented:
Yes... the lambda source needs to change there. I think we'll have to keep the lambda source within this repo too now as I think about it as we need to be able to get the hash of it [here](https://github.com/alphagov/govuk-terraform-provisioning/pull/46/files#diff-e4523e052853d909202ff683d5b738bbR50) unless we upload the source and the zip as part of the Jenkins job that pulls the lambdas in. That might be a better idea.

I think we also need to add support for the 3 environments too don't we for both the lambda source and the bucket where the emails land? Might need some pointers on that...

surminus commented:
@gpeng so in theory you could put per environment stuff within `resources/staging` or `resources/production` for example, and then when we deploy to whichever environment it'll only pull in stuff from there (and also anything in `resources/`), so env specific stuff needs to live there.

Agree it would be better to upload the source and zip as part of the Jenkins job which uploads the lambda... so just to clarify, we think it would grab the zip, upload to the bucket, get the source hash and then we can pass the source hash and location of the zip in the bucket in environmental variables to the Terraform part of the deploy process? Does that make sense?

gpeng commented:
@surminus I've updated this to get the lambda from S3 now as discussed. We need to pass in the S3 object version as per the last commit message. I think this should be pretty much good to go now. Just need to work out which repo to put the lambda source in.

surminus commented:
I'm going to write the deployment job today. The last thing to decide as you suggest is where the source goes, I'd suggest either a new repo (govuk-lambda-app-deployment?) or the govuk-app-deployment repo? We could put it in this repo but ideally I think this should be separate from Terraform code. Either way it should be easy enough to update the deployment job.

surminus commented on projects/email_alert_notifications/resources/email_alert_notifications.tf:
```diff
@@ -0,0 +1,78 @@
+  default = "govuk-email-alert-notifications"
+}
+
+variable "environment"{
```
This isn't required because it's set by default.

surminus commented on projects/email_alert_notifications/resources/email_alert_notifications.tf:
```diff
@@ -0,0 +1,78 @@
+resource "aws_lambda_function" "rename_email_files_with_request_id"{
+  s3_bucket = "${var.lambda_bucket}-${var.environment}"
+  s3_key="rename_email_files_with_request_id.zip"
+  s3_object_version="${var.rename_email_files_with_request_id_version}"
```
If I'm reading [this correctly](https://github.com/hashicorp/terraform/pull/5370) it suggests this isn't a required setting, and it would make life much easier to pull the latest by default. Makes life more difficult if we wish to deploy a specific version, though.

radeksimko replied:
Unfortunately that's not how this optional field works.
The only reason it's optional is because some people prefer to use `s3_key` for something like `rename_email_files_with_request_id-version0.2.3.zip` - i.e. leverage the object key for versioning.

Something still has to trigger the update (and generate the diff) and Lambda API won't talk to S3 directly and ask for new versions neither will Terraform (out of the box).

You would either need to [`taint`](https://www.terraform.io/docs/commands/taint.html) the `aws_lambda_function` which would recreate it from scratch and (yes) pull down the latest version from S3 or use `s3_object_version` and keep bumping the version per each release.

Or wait for the nearest Terraform release which will contain [data sources](https://github.com/hashicorp/terraform/issues/4169) + someone who will implement S3 object data source - I actually started working on it...
https://github.com/hashicorp/terraform/compare/master...TimeIncOSS:f-aws-s3-object-data-source

surminus replied:
Ah OK, I see. Thanks for clarifying! The original plan was to grab the `VersionId` using another tool and set it as an environment variable and I was just looking to cut a corner.

alexmuller commented on Rakefile:
```diff
@@ -1,6 +1,7 @@
 require 'fileutils'
 require 'rspec/core/rake_task'
 require 'tmpdir'
+require 'aws-sdk-resources'
```
Does this need to go in the `Gemfile`?

surminus replied:
The dependency is pulled in by `awspec` so I don't _think_ it also needs to be included in the Gemfile.

alexmuller replied:
Yep sorry, you're right!

alexmuller commented on Rakefile:
```diff
@@ -121,6 +122,21 @@ task apply: [:configure_state] do
+     if obj.key == lambda_filename && obj.is_latest
+       ENV['TF_VAR_LAMBDA_VERSIONID'] = obj.version_id
+     end
+   end
```
Instead of iterating over all of the `bucket.object_versions` (which might grow quite quickly) I think calling `bucket.object(lambda_filename).version_id` will return the latest version. It might also be worth guarding against the case where that value is `"null"` (which I think is what happens when versioning isn't enabled).

surminus replied:
Perfect!

alexmuller commented on Rakefile:
```diff
@@ -121,6 +122,21 @@ task apply: [:configure_state] do
 
+desc 'Latest Lambda version ID'
+task :latest_version_id do
+lambda_filename = ENV['TF_VAR_LAMBDA_FILENAME']
```
Nitpick: this could use a bit of identation.

alexmuller commented:
I'm not sure I understand the template change in 48ad5776566331d07fea4a942a986bc2982b7eb1. I think it's duplicating the templates but I'm not sure why we can use a relative path of `projects/email_alert_notifications/resources/templates/foo.json`. Going to have a play with this now.

surminus commented:
@alexmuller you're right, we could just have a full relative path in the Terraform code. I thought it was easier to read in it's own directory, but I don't really mind either way!

alexmuller commented:
Maybe something like 1cfb37562da8dd8a83404ba479d1c41c2c9de01b, is that ok? I think that's what @deanwilson had in mind with the directory structure but I might be wrong.

surminus commented:
Fine with me!

alexmuller commented:
I know hardly anything about how this SES/Lambda interaction works but the code seems like it probably makes sense. Let's run it in integration and see what happens.

boffbowsh commented:
@alexmuller :wave: I know these things. Shout if I can help.

alexmuller commented:
There's a bit of a loop here where this Terraform code creates an S3 bucket but the Rakefile relies on that bucket existing.

alexmuller commented:
@boffbowsh: Thanks!

surminus commented:
I didn't think the Rakefile should rely on the bucket existing that the Terraform code creates? It relies upon a different bucket where we upload the Lambda app, then creates a new bucket for a different function (I think).

alexmuller commented:
@surminus: Ah yep sorry, you're right. I have no idea how any of this works.
